src/diff/diffDirectories.ts

```diff
@@ -9,125 +9,102 @@ import path from "path";
- *
- * @param dir1 - One of the directories with raml files
- * @param dir2 - The other directory with raml files
- * @param commonRamls - List of all the ramls to be compared
- * @param apiCollectionChanges - Instance of ApiCollectionChanges
- * @returns An array of differences
- */
-async function diffCommonRamls(
-  dir1: string,
-  dir2: string,
-  commonRamls: string[],
-  apiCollectionChanges: ApiCollectionChanges
+export async function diffApiMetadata(
+  leftNode: ApiMetadata | undefined,
+  rightNode: ApiMetadata | undefined,
+  changes: ApiCollectionChanges
 ): Promise<void> {
-  for (const raml of commonRamls) {
-    const leftRaml = path.join(dir1, raml);
-    const rightRaml = path.join(dir2, raml);
-    try {
-      const apiDifferencer = new ApiDifferencer(leftRaml, rightRaml);
-      apiCollectionChanges.changed[
-        raml
-      ] = await apiDifferencer.findAndCategorizeChanges();
-    } catch (error) {
-      ramlToolLogger.error(`Diff operation for '${raml}' failed:`, error);
-      apiCollectionChanges.errored[raml] = error.message;
-    }
+  // If either of the nodes passed is an ApiModel object, then we have
+  // traversed that tree branch entirely and can pass the APIs for diff to
+  // be run upon
+  if (
+    (typeof leftNode === "undefined" || leftNode instanceof ApiModel) &&
+    (typeof rightNode === "undefined" || rightNode instanceof ApiModel)
+  ) {
+    return diffApiModels(leftNode, rightNode, changes);
+  }
+
+  const leftChildren = leftNode ? leftNode.children : [];
+  const rightChildren = rightNode ? rightNode.children : [];
+  // Convert the right children array into a map so that we can easily look up
+  // elements in the left array
+  const rightMap = new Map(rightChildren.map((o) => [o.name.original, o]));
+  for (const leftChild of leftChildren) {
+    const rightChild = rightMap.get(leftChild.name.original);
+    await diffApiMetadata(leftChild, rightChild, changes);
+    // Once an element is found in the right map delete it so that in the end
+    // we are only left with the elements unique to the right array
+    rightMap.delete(leftChild.name.original);
+  }
+  // Process the objects only present in the right array
+  for (const rightChild of rightMap.values()) {
+    await diffApiMetadata(undefined, rightChild, changes);
   }
 }
 
 /**
- * Finds differences between the given directories for all the raml files in the
- * provided config.
+ * Finds differences between the given directories for all the raml files.
  *
- * @param baseConfigFile - Existing APIs
- * @param newConfigFile - Newly downloaded APIs
+ * @param baseDir - Existing APIs
+ * @param newDir - Newly downloaded APIs
  *
- * @returns An array of RamlDiff objects containing differences between all the
- * old and new RAML files
+ * @returns An ApiCollectionChanges object containing all the changes
  */
 export async function diffRamlDirectories(
-  baseConfigFile: string,
-  newConfigFile: string
+  baseDir: string,
```

Comment:
Does this need a change in readme as the input is a directory instead of a config file?

Reply:
Yes, it does. I'll make the changes.